### PR TITLE
Add A* algorithm for shortest-path

### LINKED ISSUE #299: [shortest-path] astar

### PR DESCRIPTION
**Pull request by dkniffin**
While working on a project, I needed shortest-path to work faster. I noticed this library was missing A*, so I decided to add it myself. This resolves #299

Let me know if there's anything I need to change here.

### CODE REVIEW

File: src/shortest-path/astar.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+      // Skip bad paths that were enqueued before finding a better one
+      qcost = enqueued[curnode][0];
+      h = enqueued[curnode][1];
+      if (qcost < dist);
+        continue;
+    }
+
+    explored[curnode] = parent;
+    neighbors = graph.outboundNeighbors(curnode);
```
Comment (Yomguithereal) [inline src/shortest-path/astar.js:131]: callback-based methods are usually faster (e.g. `forEachOutboundNeighbor`)

File: src/shortest-path/astar.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    for (i = 0; i < neighbors.length; i++) {
+      neighbor = neighbors[i]
+      edge = graph.edge(curnode, neighbor);
```
Comment (Yomguithereal) [inline src/shortest-path/astar.js:135]: Avoid lookups for edge ids if you can (if you iterate over edges rather than neighbors, you get the necessary ids in the callback arguments in principle).

File: src/shortest-path/astar.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      cost = getEdgeWeight(edge, graph.getEdgeAttributes(edge));
+
+      if (cost == null)
+        continue
+
+      ncost = dist + cost
+
+      if (Object.entries(enqueued).includes(neighbor)) {
```
Comment (Yomguithereal) [inline src/shortest-path/astar.js:143]: Idem

File: src/shortest-path/test/astar.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  ['6', '7', 5],
+  ['7', '8', 4],
+  ['7', '9', 1],
+  ['8', '9', 10],
+  ['10', '11', 3]
+];
+
+var heuristic = (a, b) => 0
+
+describe('astar', function () {
```
Comment (Yomguithereal) [inline src/shortest-path/test/astar.js:28]: Did you test the case where a proper heuristic is given?

**Review by Yomguithereal — CHANGES_REQUESTED** [review]
Hello @dkniffin. Thank you very much for your PR. Here is my review.

File: src/shortest-path/astar.d.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  NodeAttributes extends Attributes = Attributes,
+  EdgeAttributes extends Attributes = Attributes
+>(
+  graph: Graph<NodeAttributes, EdgeAttributes>,
+  source: unknown,
+  target: unknown,
+  getEdgeWeight?:
+    | keyof EdgeAttributes
+    | MinimalEdgeMapper<number, EdgeAttributes>,
+  heuristic?: (node: unknown, target) => Number,
```
[the remark this answers is not in the document]
Comment (dkniffin) [reply src/shortest-path/astar.d.ts:14]: Ah yeah. I set these up after the fact, and I don't use Typescript much these days. I'll fix that 👍

File: src/shortest-path/astar.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  require('graphology-utils/getters').createEdgeWeightGetter;
+var Heap = require('mnemonist/heap');
+
+/**
+ * Defaults & helpers.
+ */
+var DEFAULT_WEIGHT_ATTRIBUTE = 'weight';
+
+function ASTAR_HEAP_COMPARATOR(a, b) {
+  if (a[0] > b[0]) return 1;
```
[the remark this answers is not in the document]
Comment (dkniffin) [reply src/shortest-path/astar.js:18]: tbh, I'm not 100% sure. I copied this from networkx and they just had the standard heap pop/push operators, I think. Here's their code: https://github.com/networkx/networkx/blob/main/networkx/algorithms/shortest_paths/astar.py#L103-L104

In the A* algorithm, the highest proprity should be the element that has the lowest weight + the value of the heuristic. In the way it's implemented here, that's the first element of the array. However, I don't know if there should be a "backup" priority. Maybe just the weight 🤷‍♂️

File: src/shortest-path/astar.js [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      node = parent;
+      while (node != null) {
+        path.push(node);
+        node = explored[node];
+      }
+      path.reverse();
+      return path;
+    }
+
+    if (Object.keys(explored).includes(curnode)) {
```
[the remark this answers is not in the document]
Comment (dkniffin) [reply src/shortest-path/astar.js:118]: Ah, great point. The python equivalent was `if curnode in explored`, but I wasn't sure what to translate that into in JS. I don't use `if ... in` much. In fact, I didn't even realize JS had that 😅 I'll swap it out 👍

File: src/shortest-path/test/astar.js (hunk 4 above)
Comment (dkniffin) [reply src/shortest-path/test/astar.js:28]: No. I wasn't really sure how to do that. I guess the way to test that would be to check which paths were explored. But I don't know how to do that.

Alternatively, I could give this a heuristic as a "smoke test", and check that it still finds the shortest path, but not actually test whether it made more efficient.

Comment (Yomguithereal) [reply src/shortest-path/test/astar.js:28]: I am not sure I remember well but isn't a A* pathfinding without relevant heuristic just the same as Dijkstra? In which case, don't you use a proper heuristic when you use it? Else why would it be faster than Dijkstra?

File: src/shortest-path/astar.js (hunk 1 above)
Comment (dkniffin) [reply src/shortest-path/astar.js:131]: I tried this and ESLint started erroring with `Function declared in a loop contains unsafe references to variable(s)`. I think this needs to stay as a `for` loop. But I'll switch to using edges.

File: src/shortest-path/astar.js (hunk 3 above)
Comment (dkniffin) [reply src/shortest-path/astar.js]: In this case, I want the values. 🤔 I switched to `Object.values`, but is there a better way to do that?

File: src/shortest-path/test/astar.js (hunk 4 above)
Comment (Yomguithereal) [reply src/shortest-path/test/astar.js:29]: I understand why it could be difficult to test that A* is "working" without peeking in its insides indeed. However, a paranoid test with a nontrivial heuristic function could be nice, such as the one networkx uses: https://github.com/networkx/networkx/blob/main/networkx/algorithms/shortest_paths/tests/test_astar.py

Maybe it is also possible to design a test where there exists two equivalent path that dijkstra would choose arbitrarily but that the A* heuristic forces to choose, but it might be a hassle and I would not refuse the PR in the absence of such a test in any case.

My confusion came from this statement of yours:

> While working on a project, I needed shortest-path to work faster.

I guess my question was more about your particular use-case where you actually needed A* to compute shortest paths faster and I was befuddled when your answer was:

> No. I wasn't really sure how to do that. I guess the way to test that would be to check which paths were explored. But I don't know how to do that.

because I did not understand it was related to the "testability" of the A* mechanism, as you demonstrated further afterwards. Sorry for my misunderstanding here.

File: src/shortest-path/astar.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Yomguithereal) [reply src/shortest-path/astar.js:18]: The python code you show will consider the ordering of the tuples inserted into the heap array. This means comparison will be done on the first item, then the second, then the first etc. in case of ties. So I think you will need to refine your comparator to reflect this. What's more, the final item is often the node key or some node index to break ties deterministically.

In networkx code, the heap ordering seems to only happen on some cost, then an incremental id to break ties: https://github.com/networkx/networkx/blob/main/networkx/algorithms/shortest_paths/astar.py#L115, which means that items 2, 3 and 4 are not used for sorting indeed.

In our case, this probably mean you must update the comparator to use the second element of the item to break up ties.

File: src/shortest-path/astar.js (hunk 1 above)
Comment (Yomguithereal) [reply src/shortest-path/astar.js]: The solution is to declare the functions before using them in a loop. This can make the code a little bit hard to follow because of closure state and spaghetti-ness though. If you don't feel confident with this, don't worry and keep a for loop, I will refactor into callback code later (it is more performant indeed, even more so that iterators).

Comment (Yomguithereal) [reply src/shortest-path/astar.js]: Here is an example of what I mean: https://github.com/graphology/graphology/blob/master/src/traversal/bfs.js#L59-L78

File: src/shortest-path/astar.js (hunk 3 above)
Comment (Yomguithereal) [reply src/shortest-path/astar.js]: If you need to test a value you need to perform an O(n) iteration on the object anyway indeed. You can save up an allocation by writing a `for ... in ` loop though. Is the python code making such a O(n) lookup? This looks suboptimal.

File: src/shortest-path/test/astar.js (hunk 4 above)
Comment (dkniffin) [reply src/shortest-path/test/astar.js:29]: Ah. Yes, I "tested" it in my own code (outside this library). In my use-case, I am using a graph of streets from OpenStreetMap, and I want to find the shortest path of >100k point pairs. I used a heuristic of euclidean distance, and it was WAY faster to use A* than Dijkstra. I didn't measure before/after, but it was something like several minutes before, and a few seconds after.

File: src/shortest-path/astar.js (hunk 1 above)
Comment (dkniffin) [reply src/shortest-path/astar.js]: Okay. I think I'd rather you do that refactor then, if that's alright.

Comment (Yomguithereal) [reply src/shortest-path/astar.js]: Fair enough :)

**Comment by dkniffin** [comment]
@Yomguithereal Is there anything else you need me to do on this before it gets merged?

**Comment by dkniffin** [comment]
@Yomguithereal good catch. Yes, that should be the key, not the value. I just fixed it

**Comment by Yomguithereal** [comment]
Don’t worry, I **will** merge this at some point. I just don’t have time right now.

**Comment by dkniffin** [comment]
@Yomguithereal Okay, no worries. I _think_ I spoke too soon on this change speeding up my code significantly. I think there was a bug and now it's slow again 😓 But I think this is a good change to have in the library regardless. I'm sure others would like to have A* available.

**Comment by Yomguithereal** [comment]
@dkniffin will merge now and will release soon

**Comment by Yomguithereal** [comment]
`graphology-shortest-path` v2.1.0 is now live on npm
